src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs

```diff
@@ -43,7 +43,9 @@ public class Dictionary<TKey, TValue> : IDictionary<TKey, TValue>, IDictionary,
 
         public Dictionary(IEqualityComparer<TKey>? comparer) : this(0, comparer) { }
 
-        public Dictionary(int capacity, IEqualityComparer<TKey>? comparer)
+        public Dictionary(int capacity, IEqualityComparer<TKey>? comparer) : this(0, comparer, false) { }
+
+        internal Dictionary(int capacity, IEqualityComparer<TKey>? comparer, bool useProvidedComparer)
```

Comment:
This method is going to be stamped many times for each Dictionary instantiation. It may help a tiny bit in startup of a trivial application, but it will hurt more startup of real world applications that typically have many Dictionary instantiations.

Reply:
That was my concern yeah. Is there a better way to bypass the nonrandomized initialization? I couldn't think of one.

Reply:
It depends on how far you want to go. There is no way to do that via Dictionary public APIs by design. You can always write your own custom collection that is optimized for the specific case. It can be even implemented in C in Mono VM - it would drop the number of managed methods executed by AppContext to zero. Of course, this approach gets very message fast.

I think it only makes sense to do these sorts of optimizations when they are general goodness. For example, you can make the initialization of the non-randomized comparers lazy that would general goodness that would address some of the methods you are trying to eliminate. Something like: https://github.com/dotnet/runtime/commit/48c4e31a504a39f7f59bec9dcdf9c605c65a34fb

We have been on a general plan to write more in C#. These sorts of optimizations are sometimes at odds with this plan. For example, it would be preferable to do the UTF8 -> UTF16 conversion of the appcontext keys in C# to be inline with the general plan. Of course, it would introduce more managed methods that you are worried about. Have you looked into what it would take to AOT pre-compile the methods on the startup path for Blazor wasm so that we do not need to worry about every method executed?

Reply:
Making the comparers lazy was an option that occurred to me, so I may look into doing that instead if it's more generally beneficial. I hadn't thought of doing it the way you did it in that commit, so thanks for pointing it out! The dictionary stuff is at least the lighter part of the dependencies in appcontext startup. In the process of working on this PR I discovered a way to make static cctors cheaper, which should help for this case too (the majority of the code being pulled in is static constructors and instance constructors).

For AOT our startup profile has a very different balance of where the time goes, but we're still in a situation where not every browser scenario can be AOT'd yet, so I've been trying to look for hotspots in both execution modes and identify improvements that are either easy or significant, since both interpreted and AOT are in a thousand-cuts situation right now where most of the hotspots are in the 0.5%-2.0% range.

I think doing the UTF8->UTF16 conversion for appcontext keys in C# isn't unreasonable; in this case my immediate reaction was 'why are we doing strlen in here when we already know the length?' followed by 'wow, wcslen brings in ~1200 bytes of IL by itself'. File size does become a question, i.e. is it truly beneficial to migrate the UTF8->UTF16 conversion from C into C# if the C version of those helpers is still in the binary alongside the C# one? Ideally, we could eventually migrate everything and then we don't need the C encoding helpers anymore.

To provide some context, in some older interpreted startup measurements, 7.82% of cpu cycles were spent under appctx initialization, and with these changes it went down to 3.8%. So it's not dramatic on a level that would justify making BCL code worse, but it's measurable enough that I'd want to push through on the change to the Setup method in AppContext.